### PR TITLE
Add features needed by Replatforming.

### PR DESCRIPTION
**Pull request by sengi**
We need some more flexibility in how Plek constructs URLs for services, in order to keep the complexity of govuk-helm-charts within reasonable bounds and keep everything maintainable.

- Support empty domain suffixes. This simplifies GOV.UK's Helm charts by eliminating the need to construct domain names containing the environment name. We can simply use the cluster DNS like we're supposed to, so the hostname for Content Store is just `content-store` and the local resolver expands that to `content-store.apps.svc.cluster.local`. This makes it somewhat easier to retire Plek in the long run.
- Support a list of hostnames to exclude from `PLEK_HOSTNAME_PREFIX`. This simplifies the configuration of the draft stack by saving us from having to specify a bunch of overrides for every draft app.
- Support constructing `http://` URLs for within-cluster services. (This traffic stays within the VPC, so it's encrypted before it hits the optical network anyway. The old GOV.UK doesn't actually use TLS between the load balancer and the EC2 VMs, so this is equivalent to what we already have.)

Also refactor `find()` to make the logic easier to follow.

This change is intended to be backward compatible with all current usage in govuk-puppet and client apps. In other words it's only supposed to affect Replatforming and not the existing setup.

Tested: added tests to cover the new functionality.

#### Alternatives considered

Previously our intent was to configure every backend address explicitly for every client app, with a view to eventually retiring Plek. Unfortunately that turned out to be much less feasible that any of us had anticipated, particularly when we consider how to configure the draft stack while keeping the Helm charts maintainable (avoiding excessive or risky duplication without having a silly number of levels of templating) and guarding against the most undesirable failure modes in the event of configuration errors (like ensuring it's still really hard to accidentally have draft content go live before it's supposed to).

Some more of the background to this is outlined in https://github.com/alphagov/govuk-helm-charts/commit/b7e360b. The preferred approach proposed in this PR will avoid a lot of that sort of complexity.

### CODE REVIEW

**Review by nsabri1 — APPROVED** [review]
LGTM - logically - but maybe better for someone with Ruby background

**Comment by sengi** [comment]
> LGTM - logically - but maybe better for someone with Ruby background

Thanks for the speedy review! Yup definitely could use a review for Ruby style and general concept/design.

Hopefully this isn't too contentious though — it should be 100% back-compatible and landing it soon will unblock and simplify a ton of stuff for Replatforming.

File: test/plek_test.rb
Comment (kevindew) [inline test/plek_test.rb:161]: I don't think we should add this test as it doesn't execute a different unit of code. We should probably have one that tests that a NoConfiguration exception occurs if GOVUK_APP_DOMAIN raises an error isn't set, but since you didn't add the code I don't think it's necessary you add one (it already has coverage in asset_root and website_root test files)

File: lib/plek.rb
Comment (kevindew) [inline lib/plek.rb:175]: How come we need this? I'd guess we've got a problem that the URLs don't work if they have a trailing period?

File: lib/plek.rb
Comment (kevindew) [inline lib/plek.rb:177]: I'm a bit confused by these lines, are we using `.` as a wildcard? that only applies if you have an empty suffix? I'd wonder if this is a bit cryptic as I didn't catch this until reviewing the tests.

I wonder if there are some more explicit options that can be explored. We could make an assumption that hostname with only a single label is always HTTP as that's what we anticipate for now, or could have an env var of `PLEK_HTTP_SINGLE_LABEL` hostname perhaps that's a toggle?

File: lib/plek.rb
Comment (kevindew) [reply lib/plek.rb:43]: Also on this one, now I think I understand about the special `.` PLEK_HTTP_DOMAINS value. Do we expect this will be used in replatforming beyond that case? I'm just wondering if we're making this a bit complex if the value will be always a `.`.

**Review by kevindew — COMMENTED** [review]
Nice work @sengi, I've made some suggestions and asked some questions. 

There's also some further documentation that'd be needed, there's the Changelog and there's also a reference to env vars in the readme: https://github.com/alphagov/plek#technical-documentation

**Comment by sengi** [comment]
I'll clean up the commits once approved — leaving them for now so it's easier to see what's changed.

File: CHANGELOG.md
Comment (kevindew) [inline CHANGELOG.md:5]: do you mean "to an"?

File: CHANGELOG.md
Comment (kevindew) [inline CHANGELOG.md:20]: Was there actually a change make for this in the end?

**Comment by sengi** [comment]
RFAL

File: lib/plek.rb
Comment (kevindew) [inline lib/plek.rb:42]: I don't think this is needed as a todo as it seems unlikely, it feels more of an implementation note you may want to leave in the commit message?

File: lib/plek.rb
Comment (kevindew) [inline lib/plek.rb:43]: Did you want to run uniq on this? I guess it's not a problem if it has duplicate entries?

File: lib/plek.rb
Comment (kevindew) [inline lib/plek.rb:82]: Super minor suggestion, so only apply if you're already in the code, but I wonder if this would readable without an extra method.

e.g 

```suggestion
    hostname = "#{host_prefix}#{hostname}" unless unprefixable_hosts.include?(hostname)
```

**Comment by sengi** [comment]
Damn I forgot about the env vars bit in the README. Just a sec :)

**Review by kevindew — COMMENTED** [review]
Quick, half finished review as you're making amends. Let me know when you're ready again.

File: CHANGELOG.md
Comment (sengi) [reply CHANGELOG.md]: Ooh good catch — I think there's no longer anything changing in that regard, other than adding a test. Deleted.

File: lib/plek.rb
Comment (sengi) [reply lib/plek.rb:41]: I thought about it but it seems like a waste of cycles and doesn't help readability, so gave it a miss.

File: lib/plek.rb
Comment (sengi) [reply lib/plek.rb]: Done.

File: lib/plek.rb
Comment (sengi) [reply lib/plek.rb:166]: Cool. I'll leave this alone until the next major version then.

**Review by sengi — COMMENTED** [review]
RFAL

**Review by kevindew — APPROVED** [review]
Nice one, looks good

**Comment by sengi** [comment]
Force-push for squash (no code changes).

**Comment by sengi** [comment]
Thanks for thorough review and all the Ruby tips!

**Pull request merged**
